src/types/CodepointWidthDetector.cpp

```diff
@@ -4,398 +4,1028 @@
-        UnicodeRange{ 0x1f3f8, 0x1f43e, 0 },
-        UnicodeRange{ 0x1f440, 0x1f440, 0 },
-        UnicodeRange{ 0x1f442, 0x1f4fc, 0 },
-        UnicodeRange{ 0x1f4ff, 0x1f53d, 0 },
-        UnicodeRange{ 0x1f54b, 0x1f54e, 0 },
-        UnicodeRange{ 0x1f550, 0x1f567, 0 },
-        UnicodeRange{ 0x1f57a, 0x1f57a, 0 },
-        UnicodeRange{ 0x1f595, 0x1f596, 0 },
-        UnicodeRange{ 0x1f5a4, 0x1f5a4, 0 },
-        UnicodeRange{ 0x1f5fb, 0x1f64f, 0 },
-        UnicodeRange{ 0x1f680, 0x1f6c5, 0 },
-        UnicodeRange{ 0x1f6cc, 0x1f6cc, 0 },
-        UnicodeRange{ 0x1f6d0, 0x1f6d2, 0 },
-        UnicodeRange{ 0x1f6d5, 0x1f6d7, 0 },
-        UnicodeRange{ 0x1f6dc, 0x1f6df, 0 },
-        UnicodeRange{ 0x1f6eb, 0x1f6ec, 0 },
-        UnicodeRange{ 0x1f6f4, 0x1f6fc, 0 },
-        UnicodeRange{ 0x1f7e0, 0x1f7eb, 0 },
-        UnicodeRange{ 0x1f7f0, 0x1f7f0, 0 },
-        UnicodeRange{ 0x1f90c, 0x1f93a, 0 },
-        UnicodeRange{ 0x1f93c, 0x1f945, 0 },
-        UnicodeRange{ 0x1f947, 0x1f9ff, 0 },
-        UnicodeRange{ 0x1fa70, 0x1fa7c, 0 },
-        UnicodeRange{ 0x1fa80, 0x1fa88, 0 },
-        UnicodeRange{ 0x1fa90, 0x1fabd, 0 },
-        UnicodeRange{ 0x1fabf, 0x1fac5, 0 },
-        UnicodeRange{ 0x1face, 0x1fadb, 0 },
-        UnicodeRange{ 0x1fae0, 0x1fae8, 0 },
-        UnicodeRange{ 0x1faf0, 0x1faf8, 0 },
-        UnicodeRange{ 0x20000, 0x2fffd, 0 },
-        UnicodeRange{ 0x30000, 0x3fffd, 0 },
-        UnicodeRange{ 0xe0100, 0xe01ef, 1 },
-        UnicodeRange{ 0xf0000, 0xffffd, 1 },
-        UnicodeRange{ 0x100000, 0x10fffd, 1 },
-    };
+    out = c;
+    return it;
 }
 
-// Routine Description:
-// - returns the width type of codepoint as fast as we can by using quick lookup table and fallback cache.
-// Arguments:
-// - glyph - the utf16 encoded codepoint to search for
-// Return Value:
-// - the width type of the codepoint
-CodepointWidth CodepointWidthDetector::GetWidth(const std::wstring_view& glyph) noexcept
+static CodepointWidthDetector s_codepointWidthDetector;
+
+CodepointWidthDetector& CodepointWidthDetector::Singleton() noexcept
+{
+    return s_codepointWidthDetector;
+}
+
+bool CodepointWidthDetector::GraphemeNext(GraphemeState& s, const std::wstring_view& str) noexcept
 {
-    char32_t codepoint = 0;
+    const auto beg = str.data();
+    const auto end = beg + str.size();
+    auto clusterBeg = s.beg + s.len;
+
+    // If it's a new string argument, we'll restart at the new string's beginning.
+    if (clusterBeg < beg || clusterBeg > end)
+    {
+        clusterBeg = beg;
+    }
+
+    if (clusterBeg >= end)
+    {
+        return false;
+    }
+
+    if (_mode != TextMeasurementMode::Graphemes)
+    {
+        return _graphemeNextWcswidth(s, end, clusterBeg);
+    }
+
+    auto clusterEnd = clusterBeg;
+    int state = s._state;
+    int width = 0;
+    int lead;
+    char32_t cp;
 
-    switch (glyph.size())
+    // The _state is stored ~flipped, so that we can differentiate

```

Comment:
this sounds like a boolean with more steps?

Reply:
Yeah I wanted the struct to be zero-initializable with Minsk overhead so that the calls have minimal overhead. Initializing the member to ~0 = 255 by default would break, as would adding a bool member (adds padding = bad).

There are ways to make all of this categorically better but I don't think it's time for that yet. First we need to get rid of the IsGlyphWide glue. :)